### PR TITLE
Disable EnvironmentModifications stack tracing

### PR DESCRIPTION
**Pull request by haampie**
TL;DR only track EnvironmentModifications origins under --debug.

Currently Spack keeps track of the origin in the code of any
modification to the environment variables. This is very slow. In fact
it's enabled unconditionally even in code paths where the origin of the
modification is never queried.

The only place where we inspect the origins of environment modifications
is before we start a build, IF there's an override of the type
`e.set(...)` after incremental changes like `e.append_path(..)`, which
is a "suspicious" change.

This is very rare though.

If an override like this ever happens, it might mean a package is
broken. If that leads to build errors, we can just ask the user to run
`spack -d install ...` and check the warnings issued by Spack to find
the origins of the problem.

---

The upside for me is that this cuts down running spack unit-test by a minute.

### CODE REVIEW

File: lib/spack/spack/util/environment.py [hunk 1]
```diff
@@ -365,14 +369,17 @@ class EnvironmentModifications(object):
...
         """Initializes a new instance, copying commands from 'other'
         if it is not None.
 
         Args:
             other (EnvironmentModifications): list of environment modifications
                 to be extended (optional)
+            traced (bool): enable or disable stack trace inspection to log the origin
+                of the environment modifications.
         """
+        self.traced = tracing_enabled if traced is None else bool(traced)
```
Comment (alalazo) [inline lib/spack/spack/util/environment.py:382]:
```suggestion
        self.traced = traced or tracing_enabled
```

**Review by alalazo — CHANGES_REQUESTED** [review]
LGTM. One nitpick.

File: lib/spack/spack/util/environment.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/util/environment.py:382]: So, traced = None should correspond with "defaults", otherwise traced is a bool. If I got with your suggestion you can't pass False?

Comment (alalazo) [reply lib/spack/spack/util/environment.py:382]: Nevermind, I overlooked that we need to distinguish `False` from `None` in this context.

**Review by alalazo — APPROVED** [review]
